src/Common/src/CoreWCF/Runtime/AsyncLock.cs

```diff
@@ -2,244 +2,155 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
-using System.Collections.Generic;
-#if DEBUG
-using System.Diagnostics;
-#endif
+using Microsoft.Extensions.ObjectPool;
 
 namespace CoreWCF.Runtime
 {
-    internal class AsyncLock
+    internal class AsyncLock : IAsyncDisposable
     {
-#if DEBUG
-        private StackTrace _lockTakenCallStack;
-        private string _lockTakenCallStackString;
-#endif
-        private readonly SemaphoreSlim _semaphore;
-        private readonly SafeSemaphoreRelease _semaphoreRelease;
-        private static readonly AsyncLocal<object> s_heldLocks = new AsyncLocal<object>(LockTakenValueChanged);
+        private static readonly ObjectPool<SemaphoreSlim> s_semaphorePool = (new DefaultObjectPoolProvider { MaximumRetained = 100 })
+            .Create(new SemaphoreSlimPooledObjectPolicy());
+
+        private AsyncLocal<SemaphoreSlim> _currentSemaphore;
+        private SemaphoreSlim _topLevelSemaphore;
+        private bool _isDisposed;
 
         public AsyncLock()
         {
-            _semaphore = new SemaphoreSlim(1);
-            _semaphoreRelease = new SafeSemaphoreRelease(this);
+            _topLevelSemaphore = s_semaphorePool.Get();
+            _currentSemaphore = new AsyncLocal<SemaphoreSlim>();
         }
 
-        private static void LockTakenValueChanged(AsyncLocalValueChangedArgs<object> obj)
+        public Task<IAsyncDisposable> TakeLockAsync()
         {
-            // Without this fixup, when completing the call to await TakeLockAsync there is
-            // a switch of Context and _heldLocks will be reset to null. This is because
-            // of leaving the task.
-
-            if (obj.ThreadContextChanged)
-            {
-                s_heldLocks.Value = obj.PreviousValue;
-            }
+            return TakeLockAsync(Timeout.InfiniteTimeSpan);
         }
 
-        public Task<IDisposable> TakeLockAsync()
+        public Task<IAsyncDisposable> TakeLockAsync(TimeSpan timeSpan)
         {
```

Comment:
You've added a TimeSpan variant here, but unfortunately that's not supportable. The only way this might be supportable is to go one step further and implement a custom awaiter. The problem is that any timeouts will return a faulted Task containing an exception. This means we never return the IAsyncDisposable to the using block, and we can't put back the original _currentSemaphore in the correct ExecutionContext. Whenever an async method is called, any code executed after an await is running in a copy of the ExecutionContext and no changes to an AsyncLocal will be reflected in the callers ExecutionContext. So returning from an async method effectively undoes any AsyncLocal changes. This means once we've awaited currentSemaphore.WaitAsync(), any changes we make to an AsyncLocal won't be applied in the ExecutionContext of the calling code, even if we catch the exception, modify the AsyncLocal then rethrow.  

This is why SafeSemaphoreRelease.DisposeAsync isn't an async method, it does all the AsyncLocal book keeping first, then returns the ValueTask from DisposeCoreAsync which is an async method. This ensures the book keeping which modifies an AsyncLocal happens on the callers ExecutionContext. As I mentioned, a way around this might be to create a custom awaiter which modifies things before the await continuation callback is called. But I'm not sure if even that will work as I believe the callback uses it's own ExecutionContext to run the continuation. But I might be wrong on that.  

Anyway, the result is that this AsyncLocal implementation can't implement a timeout mechanism. I think this is fine as it's targeted at replacing lock(_thisLock) which doesn't have a timeout mechanism anyway (I know Monitor.Enter does, but that's not what's been targeted here).

Reply:
[According to the docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.Threading.SemaphoreSlim.WaitAsync);k(SolutionItemsProject);k(DevLang-csharp)%26rd%3Dtrue&view=net-5.0) timeouts should rather result in a task containing false and not an exception. But reading again through my changes they suffer from some different issues. 

To avoid any risks I removed the timeout overload and simply used SemaphoreSlim in `InstanceProviderTests.cs`. I tried to keep the logic close to the original code that used `AsyncLock`. It feels also cleaner because we're not linking the AsyncLock.cs class anymore into the test project.